### PR TITLE
feat(core): improve usage filtering in recursive symbol resolution

### PR DESCRIPTION
**Pull request by kopekC**
# Motivation

# Content

The change enhances the usages method to better handle recursive symbol resolution by:
- Filtering out invalid matches during recursion
- Only including direct function calls and class instantiations
- Ensuring proper handling of chained method calls
- Preventing duplicate usages through seen set tracking

# Testing

# Please check the following before marking your PR as ready for review

- [ x] I have added tests for my changes
- [ x] I have updated the documentation or added new documentation as needed

### CODE REVIEW

**Comment by codegen-team** [comment]
### Technical Review Summary

This PR enhances the `usages` method by adding depth-based usage traversal capabilities. Key changes:

1. New `max_depth` parameter to control usage traversal depth
2. Automatic usage type inference based on depth
3. Comprehensive test coverage with various scenarios

The implementation is well-structured and maintains backward compatibility.

**Comment by codegen-team** [comment]
### Detailed Review

**Strengths:**
1. Smart default behavior with usage type inference based on depth
2. Well-implemented recursive collection for chained usages
3. Comprehensive test suite covering basic scenarios, edge cases, and complex patterns

**Suggestions:**
1. Consider adding a safeguard against circular dependencies in the recursive collection
2. Documentation is clear but could benefit from an example showing depth-based traversal

Overall, the implementation is solid and the feature will be valuable for more granular usage analysis.

**Comment by codegen-team** [comment]
Technical Review Summary:

Strengths:
1. Well-structured implementation of depth-based traversal
2. Comprehensive test coverage with edge cases
3. Maintains backward compatibility
4. Clear documentation of the new parameter

Areas for Improvement:
1. Consider adding a maximum depth limit to prevent potential stack overflow in deeply nested usage chains
2. The seen set could potentially grow large for big codebases - consider adding a size limit or warning
3. The recursive collection of usages might benefit from being extracted into a separate private method for better maintainability

The implementation is solid and ready for production, but these suggestions could make it even more robust.

**Comment by codegen-team** [comment]
Detailed Technical Review:

Implementation Analysis:
1. Core Changes:
   - Added `max_depth` parameter to `usages` method
   - Implemented recursive traversal for chained usages
   - Added intelligent default usage type selection based on depth
   - Maintained backward compatibility

2. Code Quality:
   - Clean implementation of the depth-based traversal
   - Good use of seen set to prevent cycles
   - Clear parameter documentation
   - Logical organization of the recursive collection logic

3. Test Coverage:
   - Comprehensive test suite with three distinct test files
   - Good coverage of basic usage, complex scenarios, and edge cases
   - Clear test organization and naming
   - Thorough validation of depth behavior

Suggestions for Improvement:

1. Implementation:
   - Consider adding a maximum depth limit (e.g., MAX_DEPTH = 10) to prevent stack overflow
   - The seen set could benefit from a size limit for large codebases
   - Consider extracting the recursive collection logic into a private method
   - Add memory usage warnings for large traversals

2. Documentation:
   - Document the default usage type behavior more prominently
   - Add performance characteristics to the docstring
   - Consider adding examples for common use cases

3. Testing:
   - Add performance tests for deep nesting scenarios
   - Consider adding stress tests for large codebases
   - Add tests for maximum depth limit once implemented

Overall, this is a solid implementation that adds valuable functionality while maintaining code quality. The suggestions above are optimizations rather than blockers.

**Comment by kopekC** [comment]
Overall this is a solid enhancement to the usage analysis capabilities. The implementation is clean and well-tested. A few suggestions for consideration:

1. Documentation could benefit from more examples showing different depth scenarios in practice
2. Consider adding performance warnings in docstring for large depth values
3. Some defensive programming could be added around depth values and circular references
4. For large codebases, a more memory-efficient implementation might be worth considering

The test coverage is comprehensive and covers important edge cases. The automatic usage type selection based on depth is a nice touch that makes the API more intuitive.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding a `max_depth` parameter to control the depth of usage traversal in the codebase. This allows for more granular control over usage analysis, particularly useful for analyzing method chains and property accesses.

### Changes
1. Core Enhancement:
   - Added `max_depth` parameter to `usages` method
   - Implemented recursive usage collection for depth > 1
   - Added intelligent default usage types based on depth:
     - depth=1: DIRECT usages only
     - depth>1: DIRECT | CHAINED usages

2. Test Coverage:
   - Comprehensive test suite added covering:
     - Basic depth traversal scenarios
     - Complex usage patterns
     - Edge cases including nested functions and multiple inheritance
   - Tests verify behavior for different depth levels and usage type combinations

### Technical Analysis

#### Strengths
- Well-structured implementation of depth-based traversal
- Clear documentation of the new parameter and its behavior
- Comprehensive test coverage with meaningful test cases
- Efficient usage of seen set to prevent duplicate traversal
- Backward compatible change

#### Areas for Improvement
1. Performance Consideration: The recursive collection for max_depth > 1 could benefit from a comment about potential performance implications for large codebases.

2. Code Structure:
```python
if max_depth is not None and max_depth > 1:
    # For max_depth > 1, recursively collect usages
    seen = set(usages_to_return)
    for usage in list(usages_to_return):  # Create a copy to iterate over
```
Consider extracting this recursive collection logic into a separate private method for better readability and maintainability.

3. Documentation: Consider adding an example in the docstring showing the difference between depth=1 and depth=2 results.

### Verdict
The implementation is solid and well-tested. The suggested improvements are minor and the PR can be merged after addressing them if desired.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding a `max_depth` parameter to control the depth of usage traversal in the codebase. This allows for more granular control over usage analysis, particularly useful for analyzing method chains and property accesses.

### Changes
1. Core Enhancement:
   - Added `max_depth` parameter to `usages` method
   - Implemented recursive usage collection for depth > 1
   - Added intelligent default usage types based on depth:
     - depth=1: DIRECT usages only
     - depth>1: DIRECT | CHAINED usages

2. Test Coverage:
   - Comprehensive test suite added covering:
     - Basic depth traversal scenarios
     - Complex usage patterns
     - Edge cases including nested functions and multiple inheritance
   - Tests verify behavior for different depth levels and usage type combinations

### Technical Analysis

#### Strengths
- Well-structured implementation of depth-based traversal
- Clear documentation of the new parameter and its behavior
- Comprehensive test coverage with meaningful test cases
- Efficient usage of seen set to prevent duplicate traversal
- Backward compatible change

#### Areas for Improvement
1. Performance Consideration: The recursive collection for max_depth > 1 could benefit from a comment about potential performance implications for large codebases.

2. Code Structure:
```python
if max_depth is not None and max_depth > 1:
    # For max_depth > 1, recursively collect usages
    seen = set(usages_to_return)
    for usage in list(usages_to_return):  # Create a copy to iterate over
```
Consider extracting this recursive collection logic into a separate private method for better readability and maintainability.

3. Documentation: Consider adding an example in the docstring showing the difference between depth=1 and depth=2 results.

### Verdict
The implementation is solid and well-tested. The suggested improvements are minor and the PR can be merged after addressing them if desired.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding a `max_depth` parameter to control the depth of usage traversal in the codebase. This is a significant improvement for controlling the granularity of usage analysis.

### Changes Overview
1. Core API Enhancement:
   - Added `max_depth` parameter to `usages()` method
   - Implemented recursive usage collection for depth > 1
   - Added intelligent default behavior for usage types based on depth

2. Test Coverage:
   - Added comprehensive test suite in `test_usable_usages.py`
   - Covers basic depth scenarios, complex usage patterns, and edge cases

### Technical Details
- When `max_depth=1`: Only direct usages are returned
- When `max_depth>1`: Includes both direct usages and chained usages (method calls and property accesses)
- Default usage types are now context-aware:
  - `max_depth=1` → `UsageType.DIRECT`
  - `max_depth>1` → `UsageType.DIRECT | UsageType.CHAINED`

### Potential Improvements
1. Performance Consideration: The recursive collection for `max_depth > 1` could benefit from a more efficient algorithm to handle large codebases. Consider adding a warning or limit for very large depth values.

2. Documentation:
   - Consider adding examples in the docstring showing the difference between various depth levels
   - Document the performance implications of using different depth values

3. Edge Cases:
   - Consider adding handling for circular references in the usage chain
   - Add explicit tests for error conditions and boundary cases (e.g., negative depth values)

4. Type Safety:
   - Consider adding runtime validation for the `max_depth` parameter to ensure it's a positive integer

The implementation is solid and well-tested, making it a valuable addition to the codebase's analysis capabilities.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding a `max_depth` parameter to control the depth of usage traversal in the codebase. This allows for more granular control over usage analysis, particularly useful for analyzing method chains and property accesses.

### Changes
1. Core Enhancement:
   - Added `max_depth` parameter to `usages` method
   - Implemented recursive usage collection for depth > 1
   - Added intelligent default usage types based on depth:
     - depth=1: DIRECT usages only
     - depth>1: DIRECT | CHAINED usages

2. Test Coverage:
   - Comprehensive test suite added covering:
     - Basic depth traversal scenarios
     - Complex usage patterns
     - Edge cases including nested functions and multiple inheritance
   - Tests verify behavior for different depth levels and usage type combinations

### Technical Analysis

#### Strengths
- Well-structured implementation of depth-based traversal
- Clear documentation of the new parameter and its behavior
- Comprehensive test coverage with meaningful test cases
- Efficient usage of seen set to prevent duplicate traversal
- Backward compatible change

#### Areas for Improvement
1. Performance Consideration: The recursive collection for max_depth > 1 could benefit from a comment about potential performance implications for large codebases.

2. Code Structure:
```python
if max_depth is not None and max_depth > 1:
    # For max_depth > 1, recursively collect usages
    seen = set(usages_to_return)
    for usage in list(usages_to_return):  # Create a copy to iterate over
```
Consider extracting this recursive collection logic into a separate private method for better readability and maintainability.

3. Documentation: Consider adding an example in the docstring showing the difference between depth=1 and depth=2 results.

### Verdict
The implementation is solid and well-tested. The suggested improvements are minor and the PR can be merged after addressing them if desired.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding a `max_depth` parameter to control the depth of usage traversal in the codebase. This is a significant improvement for controlling the granularity of usage analysis.

### Changes Overview
1. Core API Enhancement:
   - Added `max_depth` parameter to `usages()` method
   - Implemented recursive usage collection for depth > 1
   - Added intelligent default behavior for usage types based on depth

2. Test Coverage:
   - Added comprehensive test suite in `test_usable_usages.py`
   - Covers basic depth scenarios, complex usage patterns, and edge cases

### Technical Details
- When `max_depth=1`: Only direct usages are returned
- When `max_depth>1`: Includes both direct usages and chained usages (method calls and property accesses)
- Default usage types are now context-aware:
  - `max_depth=1` → `UsageType.DIRECT`
  - `max_depth>1` → `UsageType.DIRECT | UsageType.CHAINED`

### Potential Improvements
1. Performance Consideration: The recursive collection for `max_depth > 1` could benefit from a more efficient algorithm to handle large codebases. Consider adding a limit parameter to prevent excessive traversal.

2. Documentation:
   - Consider adding examples in the docstring showing the difference between various depth levels
   - Document the performance implications of using higher depth values

3. Edge Cases:
   - Consider adding handling for circular references in the usage chain
   - Add explicit tests for error conditions and boundary cases (e.g., negative depth values)

4. Type Safety:
   - Consider adding runtime validation for the `max_depth` parameter to ensure it's a positive integer

The implementation is solid and well-tested, making it a valuable addition to the codebase's analysis capabilities.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding a `max_depth` parameter to control the depth of usage traversal in the codebase. This is a significant improvement for controlling the scope of usage analysis.

### Changes
1. Core Enhancement:
   - Added `max_depth` parameter to `usages` method
   - Implemented recursive usage collection for depth > 1
   - Added intelligent default usage types based on depth:
     - depth=1 → UsageType.DIRECT
     - depth>1 → UsageType.DIRECT | UsageType.CHAINED

2. Test Coverage:
   - Comprehensive test suite added with three test cases:
     - Basic max_depth functionality
     - Complex scenarios (multiple inheritance, various usage patterns)
     - Edge cases (nested functions, multiple references)

### Technical Analysis
#### Strengths
- Well-structured implementation of depth-based traversal
- Clear separation of concerns in the usage collection logic
- Comprehensive test coverage with diverse scenarios
- Backward compatible change

#### Areas for Improvement
1. Performance Consideration:
   - The recursive collection for max_depth > 1 could benefit from a more efficient implementation to avoid potential stack overflow for deep hierarchies
   - Consider adding a warning or limit for very large depth values

2. Documentation:
   - The docstring could be more explicit about the performance implications of using larger depth values
   - Consider adding examples in the docstring showing different depth scenarios

3. Code Structure:
   - The seen set implementation could be moved to a separate helper method for better readability
   - Consider adding debug logging for complex traversal paths

### Conclusion
The PR represents a solid enhancement to the usage analysis capabilities. The implementation is well-tested and maintains backward compatibility while adding valuable functionality for more granular usage analysis.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding a `max_depth` parameter to control the depth of usage traversal in the codebase. This is a significant improvement for controlling the granularity of usage analysis.

### Changes Overview
1. Core API Enhancement:
   - Added `max_depth` parameter to `usages()` method
   - Implemented recursive usage collection for depth > 1
   - Added intelligent default behavior for usage types based on depth

2. Test Coverage:
   - Added comprehensive test suite in `test_usable_usages.py`
   - Covers basic depth scenarios, complex usage patterns, and edge cases

### Technical Details
- When `max_depth=1`: Only direct usages are returned
- When `max_depth>1`: Includes both direct usages and chained usages (method calls and property accesses)
- Default usage types are now context-aware:
  - `max_depth=1` → `UsageType.DIRECT`
  - `max_depth>1` → `UsageType.DIRECT | UsageType.CHAINED`

### Potential Improvements
1. Performance Consideration: The recursive collection for `max_depth > 1` could benefit from a more efficient implementation to avoid potential stack overflow for deeply nested usage chains. Consider adding a warning or limit for very large depth values.

2. Documentation: While the docstring explains the behavior, it would be helpful to add:
   - Examples showing different depth scenarios
   - Performance implications for different depth values
   - Best practices for choosing appropriate depth values

3. Edge Cases: Consider adding explicit handling for circular references in the usage chain to prevent infinite recursion.

The implementation is solid and well-tested, making it a valuable addition to the codebase's analysis capabilities.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding a `max_depth` parameter to control the depth of usage traversal in the codebase. This is a significant improvement for controlling the scope of usage analysis.

### Changes Overview
1. Core API Enhancement:
   - Added `max_depth` parameter to `usages` method
   - Implemented recursive usage collection for depth > 1
   - Added intelligent default behavior for usage types based on depth

2. Test Coverage:
   - Added comprehensive test suite in `test_usable_usages.py`
   - Covers basic depth scenarios, complex usage patterns, and edge cases

### Technical Details
- When `max_depth=1`: Only direct usages are returned
- When `max_depth>1`: Includes both direct usages and chained usages (method calls and property accesses)
- Default usage types are now context-aware:
  - `max_depth=1` → `UsageType.DIRECT`
  - `max_depth>1` → `UsageType.DIRECT | UsageType.CHAINED`

### Potential Improvements
1. Performance Consideration: The recursive collection for `max_depth > 1` could benefit from a more efficient implementation to avoid potential stack overflow for deep hierarchies. Consider adding a depth limit safeguard.

2. Documentation: While the docstring is good, it could be enhanced with:
   - Example usage patterns
   - Performance implications for different depth values
   - Best practices for choosing appropriate depth values

3. Edge Cases: Consider adding explicit handling for circular references in the usage chain to prevent infinite recursion.

The implementation is solid and well-tested, making it a valuable addition to the codebase.

**Comment by codegen-team** [comment]
Overall Review:

Technical Implementation:
- The implementation is well-structured and adds valuable functionality for controlling usage analysis depth
- Good use of caching and optimization with the seen set to prevent duplicate processing
- Comprehensive test coverage with various edge cases

Suggestions for Improvement:
1. Documentation:
   - Consider adding type hint annotations that reflect the default behavior
   - Add performance implications warning for large max_depth values
   - Document the maximum recommended depth value

2. Implementation:
   - Consider adding a maximum depth limit to prevent excessive recursion
   - The seen set implementation could potentially be moved to a separate method for better reusability
   - Consider adding performance tests for large codebases

3. Testing:
   - While the test coverage is good, consider adding performance benchmarks
   - Add tests for error cases (e.g., negative max_depth values)
   - Consider adding tests for circular dependencies

The code is production-ready but could benefit from these minor enhancements. The test suite is particularly thorough and covers most edge cases effectively.

**Comment by codegen-team** [comment]
Thank you for this well-structured PR. The implementation looks solid with comprehensive test coverage.

A few suggestions for consideration:

1. Performance: Consider adding a warning or limit for large depth values to prevent potential performance issues in large codebases.

2. Documentation: The docstring could benefit from a concrete example showing the difference between depth levels, making it more intuitive for users.

3. Type Safety: Consider adding validation for negative max_depth values to prevent unexpected behavior.

The test coverage is particularly thorough, covering basic scenarios, complex cases, and edge conditions. The implementation is clean and maintains backward compatibility.

Overall, this is a valuable addition that will make usage analysis more flexible and powerful.

**Comment by codegen-team** [comment]
## Technical Review

### Changes Overview
The PR enhances the `usages` method by adding depth-based traversal capabilities, allowing for more granular control over usage collection in the codebase.

### Implementation Analysis

#### Positive Aspects
1. Well-structured implementation of recursive usage collection
2. Comprehensive test coverage with multiple scenarios
3. Clear documentation of the new functionality
4. Smart default behavior for usage types based on depth

#### Technical Details
- The `max_depth` parameter controls usage traversal:
  - `depth=1`: Only direct usages
  - `depth>1`: Direct usages plus chained usages
- Implements cycle detection through `seen` set
- Maintains backward compatibility with existing usage type filtering

### Test Coverage
The new test file `test_usable_usages.py` is thorough and covers:
- Basic depth-based traversal
- Complex scenarios with multiple inheritance
- Edge cases including nested functions and multiple references
- Various usage type combinations

### Suggestions for Improvement

1. **Performance Consideration**: For large codebases, it might be worth adding a warning or assertion for very large depth values to prevent potential performance issues.

2. **Documentation**: Consider adding an example in the docstring showing the difference between depth=1 and depth=2 results.

3. **Type Hints**: The return type annotation could be more specific about the ordering guarantee (currently only mentioned in the docstring).

Overall, this is a solid enhancement that adds valuable functionality while maintaining code quality and backward compatibility.

**Comment by codegen-team** [comment]
## Technical Review

### Changes Overview
The PR enhances the `usages` method by adding depth-based traversal capabilities, allowing for more granular control over usage collection. This is particularly useful for analyzing code dependencies and understanding usage patterns at different levels of depth.

### Implementation Analysis

#### Positive Aspects
1. Well-structured implementation of recursive usage collection
2. Comprehensive test coverage with multiple scenarios
3. Clear documentation of the new functionality
4. Smart default behavior for usage types based on depth
5. Efficient handling of duplicate usages through the `seen` set

#### Technical Details
- The `max_depth` parameter controls traversal depth:
  - depth=1: Only direct usages
  - depth>1: Includes chained usages (method calls and property accesses)
- Default usage types are automatically set based on depth:
  - depth=1 → UsageType.DIRECT
  - depth>1 → UsageType.DIRECT | UsageType.CHAINED

### Test Coverage
The new test file `test_usable_usages.py` provides excellent coverage with:
- Basic depth-based usage scenarios
- Complex scenarios with multiple inheritance
- Edge cases including nested functions and multiple references
- Property and method chaining scenarios

### Suggestions for Improvement
1. Consider adding a maximum depth limit to prevent potential stack overflow in highly nested usage chains
2. The recursive collection could benefit from a progress callback for large codebases
3. Consider adding debug logging for the recursive traversal to help with troubleshooting

Overall, this is a solid enhancement that adds valuable functionality while maintaining code quality and providing comprehensive test coverage.

**Comment by codegen-team** [comment]
## Technical Review

### Changes Overview
The PR enhances the `usages` method by adding depth-based traversal capabilities, allowing for more granular control over usage collection. This is particularly useful for analyzing complex dependency chains and understanding code relationships at different levels.

### Implementation Analysis

#### Positive Aspects
1. Well-structured implementation of recursive usage collection
2. Comprehensive test coverage with clear test cases
3. Good documentation of the new functionality
4. Smart default behavior for usage types based on depth

#### Technical Details
- The `max_depth` parameter controls traversal depth:
  - `depth=1`: Only direct usages
  - `depth>1`: Includes chained usages (method calls and property accesses)
- Implements cycle detection through `seen` set
- Maintains backward compatibility with existing usage type filtering

### Test Coverage
The new test file `test_usable_usages.py` provides excellent coverage with:
- Basic depth-based traversal tests
- Complex scenarios including inheritance
- Edge cases with nested functions and multiple inheritance

### Suggestions for Improvement

1. **Performance Consideration**: For large codebases, it might be worth adding a warning or limit when `max_depth` is set to high values, as the recursive collection could become expensive.

2. **Documentation**: Consider adding an example in the docstring showing the difference between various depth levels, similar to how it's demonstrated in the tests.

3. **Type Hints**: The return type annotation remains unchanged. Consider documenting that the returned list might contain nested usages when `max_depth > 1`.

Overall, this is a solid enhancement that adds valuable functionality while maintaining code quality and backward compatibility.

**Comment by codegen-team** [comment]
**Suggestion for the recursive implementation:**

In the recursive usage collection (around line 89 in usable.py), consider enhancing the cycle detection mechanism. Currently, the `seen` set tracks just the usages, but for more robust handling of circular dependencies, you could track (usage, depth) pairs. This would ensure that:

1. The same usage could be revisited at different depths if needed
2. The recursion would still terminate correctly at the specified max_depth
3. It would provide more accurate depth tracking in complex dependency graphs

Example modification:
```python
# Instead of:
seen = set(usages_to_return)

# Consider:
seen = {(usage, 1) for usage in usages_to_return}
# Then in the loop:
if (next_usage, current_depth) not in seen:
    seen.add((next_usage, current_depth))
```

This would make the depth tracking more precise and robust against complex circular dependencies.

**Comment by codegen-team** [comment]
## Technical Review

### Changes Overview
The PR enhances the usage tracking functionality by adding depth-based traversal capabilities to the `usages` method. This allows for more granular control over how deep the usage chain should be analyzed.

### Implementation Analysis

#### Core Changes
1. New `max_depth` parameter with intelligent defaults:
   - `depth=1`: Only direct usages
   - `depth>1`: Direct usages plus chained usages
   - `None`: Original behavior (all usages)

2. Smart usage type selection based on depth:
   - When `max_depth=1`: Defaults to `UsageType.DIRECT`
   - When `max_depth>1`: Defaults to `UsageType.DIRECT | UsageType.CHAINED`

3. Recursive usage collection implementation for `max_depth > 1`

### Test Coverage
The new test file `test_usable_usages.py` provides excellent coverage with:
- Basic depth-based usage scenarios
- Complex usage patterns (inheritance, decorators, nested functions)
- Edge cases and boundary conditions

### Strengths
1. Well-documented changes with clear docstring updates
2. Comprehensive test coverage
3. Backward compatible implementation
4. Efficient implementation using seen set to avoid duplicates

### Suggestions for Improvement
1. Consider adding a maximum limit for `max_depth` to prevent potential stack overflow in deeply nested usage chains
2. The recursive collection could benefit from a more explicit early exit condition when no new usages are found
3. Consider adding debug logging for the recursive traversal to help with troubleshooting

### Code Quality
The implementation is clean, well-tested, and follows the project's coding standards. The changes are focused and the new functionality is well-integrated with the existing codebase.

File: src/codegen/sdk/core/interfaces/usable.py [hunk 1]
```diff
@@ -71,11 +83,24 @@
...
+        if max_depth is not None and max_depth > 1:
+            # For max_depth > 1, recursively collect usages
+            seen = set(usages_to_return)
+            for usage in list(usages_to_return):  # Create a copy to iterate over
+                if usage.usage_symbol and isinstance(usage.usage_symbol, Usable):
+                    # When recursing, we want to find method calls and property accesses
+                    next_usages = usage.usage_symbol.usages(usage_types=UsageType.CHAINED, max_depth=max_depth - 1)
+                    for next_usage in next_usages:
+                        if next_usage not in seen:
+                            seen.add(next_usage)
```
Comment (codegen-team) [inline src/codegen/sdk/core/interfaces/usable.py:98]: The recursive collection implementation here is clean, but consider adding an early exit condition when `usages_to_return` doesn't grow after processing a level. This could improve performance for cases where no new usages are found but we haven't reached max_depth yet.

Something like:
```python
initial_size = len(usages_to_return)
# ... process next level ...
if len(usages_to_return) == initial_size:
    break  # No new usages found at this level
```

File: tests/unit/codegen/sdk/core/interfaces/test_usable_usages.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        assert len(all_usages) >= 6
+
+        # Test with depth=1 but all usage types
+        depth1_all_types = child_class.usages(max_depth=1, usage_types=None)
+        # When max_depth=1, we should get the same or fewer usages than without depth
+        assert len(depth1_all_types) <= len(all_usages)
+
+        # Test with specific usage types but no depth
+        method_usages = child_class.usages(usage_types=UsageType.DIRECT | UsageType.CHAINED)
+        assert len(method_usages) >= 4  # Direct instantiations and method calls
```
Comment (codegen-team) [inline tests/unit/codegen/sdk/core/interfaces/test_usable_usages.py:147]: The test coverage is thorough, but it might be worth adding a specific test case for cyclic references to ensure they're handled correctly. For example:

```python
class A:
    def method_a(self, b):
        b.method_b(self)

class B:
    def method_b(self, a):
        a.method_a(self)

# Create cycle
a = A()
b = B()
a.method_a(b)
```

This would help verify that the seen set properly prevents infinite recursion in such cases.

File: src/codegen/sdk/core/interfaces/usable.py [hunk 3]
```diff
@@ -71,11 +83,24 @@
~            meta_data = edge[2]
~            if meta_data.type == EdgeType.SYMBOL_USAGE:
                 usage = meta_data.usage
                 if usage_types is None or usage.usage_type in usage_types:
                     usages_to_return.append(usage)
+
```
Comment (codegen-team) [inline src/codegen/sdk/core/interfaces/usable.py:88]: Consider adding a maximum recursion depth safeguard here. While the `seen` set prevents infinite loops from cycles, there could still be very deep chains of usage that might impact performance. A configurable `MAX_RECURSION_DEPTH` constant (e.g., defaulting to 10) could help prevent potential issues in large codebases.

**Comment by codegen-team** [comment]
Overall solid implementation of recursive usage analysis. A few suggestions for consideration:

1. Performance: For large codebases, recursive traversal with high depth values could be computationally expensive. Consider:
   - Adding a `max_results` parameter to limit the result set
   - Adding a warning in the docstring about performance implications of high depth values
   
2. Documentation: The docstring could benefit from concrete examples showing:
   - What results to expect at different depths
   - Common usage patterns
   - Performance considerations

The test coverage is comprehensive and well-structured, effectively covering the main use cases and edge scenarios.

File: src/codegen/sdk/core/interfaces/usable.py [hunk 4]
```diff
@@ -37,19 +37,24 @@
...
     @proxy_property
     @reader(cache=False)
-    def usages(self, usage_types: UsageType | None = None) -> list[Usage]:
+    def usages(self, usage_types: UsageType | None = None, max_depth: int | None = None) -> list[Usage]:
         """Returns a list of usages of the exportable object.
 
         Retrieves all locations where the exportable object is used in the codebase. By default, returns all usages, such as imports or references within the same file.
+        When max_depth is provided:
+        - At depth 1: Returns only direct usages
+        - At depth > 1: Returns direct usages plus chained usages (method calls and property accesses)
```
Comment (codegen-team) [inline src/codegen/sdk/core/interfaces/usable.py:52]: Consider adding a performance note in the docstring about the computational cost of high depth values, especially for large codebases. Something like:

```python
\"\"\"
Note:
    Using high depth values on large codebases may impact performance as the method
    recursively traverses the usage graph. Consider using lower depth values or
    implementing pagination if you need to process large result sets.
\"\"\"
```

File: src/codegen/sdk/core/interfaces/usable.py (hunk 3 above)
Comment (codegen-team) [inline src/codegen/sdk/core/interfaces/usable.py:88]: The recursive collection logic here is well implemented. Consider extracting the seen set initialization and management into a helper method if this pattern is used elsewhere in the codebase. This would make the code more DRY and easier to maintain.

**Comment by codegen-team** [comment]
## Technical Review

### Summary
This PR enhances the `usages` method in the `Usable` interface by adding support for recursive usage analysis through a new `max_depth` parameter. The changes allow for more granular control over usage discovery, particularly useful for analyzing method chains and property accesses.

### Changes Overview
1. Added `max_depth` parameter to `usages` method
2. Implemented recursive usage collection for depth > 1
3. Added automatic usage type inference based on depth
4. Added comprehensive test suite covering various scenarios

### Implementation Details
- The `max_depth` parameter controls how deep to traverse the usage graph:
  - `depth=1`: Only direct usages
  - `depth>1`: Direct usages plus chained usages (method calls and property accesses)
- Added automatic usage type inference:
  - When `max_depth=1`: Defaults to `UsageType.DIRECT`
  - When `max_depth>1`: Defaults to `UsageType.DIRECT | UsageType.CHAINED`
- Implemented cycle detection using a `seen` set to prevent infinite recursion
- Maintained backward compatibility with existing usage type filtering

### Test Coverage
The new test file `test_usable_usages.py` provides excellent coverage with three main test categories:
1. Basic depth-based usage discovery
2. Complex scenarios (multiple inheritance, method chains)
3. Edge cases (nested functions, multiple references)

### Suggestions for Improvement
1. **Documentation**: Consider adding examples in the docstring showing different depth scenarios
2. **Performance**: For large codebases, might want to add a warning when `max_depth` is set high
3. **Type Hints**: Consider adding a type hint for the return value of recursive calls to maintain consistency

### Code Quality
- The implementation is clean and maintains the existing code style
- Good separation of concerns between depth handling and usage type filtering
- Efficient use of sets for duplicate prevention
- Well-structured test cases that cover various scenarios

Overall, this is a solid enhancement that adds valuable functionality while maintaining backward compatibility.

File: src/codegen/sdk/core/interfaces/usable.py [hunk 5]
```diff
@@ -37,19 +37,24 @@
...
 
     @proxy_property
     @reader(cache=False)
-    def usages(self, usage_types: UsageType | None = None) -> list[Usage]:
+    def usages(self, usage_types: UsageType | None = None, max_depth: int | None = None) -> list[Usage]:
         """Returns a list of usages of the exportable object.
 
         Retrieves all locations where the exportable object is used in the codebase. By default, returns all usages, such as imports or references within the same file.
+        When max_depth is provided:
+        - At depth 1: Returns only direct usages
```
Comment (codegen-team) [inline src/codegen/sdk/core/interfaces/usable.py:51]: Consider adding an example in the docstring showing how the max_depth parameter affects the results. Something like:

```python
# Example:
class A:
    def method_a(self): pass

class B:
    def method_b(self):
        self.a = A()
        self.a.method_a()

# Usage examples:
# b = B()
# b.method_b()  # depth=1 finds only direct usage
# b.method_b()  # depth=2 finds method_b and method_a calls
```

This would make it immediately clear to users how the depth parameter works.

File: src/codegen/sdk/core/interfaces/usable.py [hunk 6]
```diff
@@ -71,11 +83,24 @@
~            meta_data = edge[2]
~            if meta_data.type == EdgeType.SYMBOL_USAGE:
                 usage = meta_data.usage
```
Comment (codegen-team) [inline src/codegen/sdk/core/interfaces/usable.py:85]: The cycle detection using `seen` set is a good approach. However, consider adding a comment explaining why this is necessary to prevent infinite recursion in cases of circular references. This would help future maintainers understand the importance of this check.

**Comment by codegen-team** [comment]
Consider adding a warning or limit for large depth values as they could lead to exponential traversal in heavily interconnected codebases. Maybe add a MAX_DEPTH constant (e.g., 5) to prevent potential performance issues.

**Comment by codegen-team** [comment]
The docstring could benefit from a concrete example showing the difference between depth=1 and depth=2 usage collection. This would make it more immediately clear to users how the depth parameter affects the results.

**Comment by codegen-team** [comment]
While the test coverage is comprehensive, it might be worth adding a specific test case for cyclic references to ensure the `seen` set properly prevents infinite recursion. This could be particularly important for real-world usage patterns.
